code/controllers/subsystem/dbcore.dm

```diff
@@ -0,0 +1,437 @@
+	/// The SQL statement being executed with :parameter placeholders
+	var/sql
+	/// An associative list of parameters to be substituted into the statement
+	var/arguments
+
+	// Status information
+	/// Is the query currently in progress
+	var/in_progress
+	/// What was our last error, if any
+	var/last_error
+	/// What was our last activity
+	var/last_activity
+	/// When was our last activity
+	var/last_activity_time
+
+	// Output
+	/// List of all rows returned
+	var/list/list/rows
+	/// Counter of the next row to take
+	var/next_row_to_take = 1
+	/// How many rows were affected by the query
+	var/affected
+	/// ID of the last inserted row
+	var/last_insert_id
+	/// List of data values populated by NextRow()
+	var/list/item
+
+// Sets up some vars and throws it into the SS active query list
+/datum/db_query/New(connection, sql, arguments)
+	SSdbcore.active_queries[src] = TRUE
+	Activity("Created")
+	item = list()
+
+	src.connection = connection
+	src.sql = sql
+	src.arguments = arguments
+
+// Takes it out of the active query list, as well as closing it up
+/datum/db_query/Destroy()
+	Close()
+	SSdbcore.active_queries -= src
+	return ..()
+
+/datum/db_query/CanProcCall(proc_name)
+	// go away
+	return FALSE
+
+
+/**
+  * Activity Update Handler
+  *
+  * Sets the last activity text to the argument input, as well as updating the activity time
+  *
+  * Arguments:
+  * * activity - Last activity text
+  */
+/datum/db_query/proc/Activity(activity)
+	last_activity = activity
+	last_activity_time = world.time
+
+/**
+  * Wrapped for warning on execution
+  *
+  * You should use this proc when running the SQL statement. It will auto inform the user and the online admins if a query fails
+  *
+  * Arguments:
+  * * async - Are we running this query asynchronously
+  * * log_error - Do we want to log errors this creates? Disable this if you are running sensitive queries where you dont want errors logged in plain text (EG: Auth token stuff)
+  */
+/datum/db_query/proc/warn_execute(async = TRUE, log_error = TRUE)
+	. = Execute(async, log_error)
+	if(!.)
+		SSdbcore.total_errors++
+		if(usr)
+			to_chat(usr, "<span class='danger'>A SQL error occurred during this operation, please inform an admin or a coder.</span>")
+		message_admins("An SQL error has occured. Please check the server logs, with the following timestamp ID: \[[time_stamp()]]")
+
+/**
+  * Main Execution Handler
+  *
+  * Invoked by [warn_execute()]
+  * This handles query error logging, as well as invoking the actual runner
+  * Arguments:
+  * * async - Are we running this query asynchronously
+  * * log_error - Do we want to log errors this creates? Disable this if you are running sensitive queries where you dont want errors logged in plain text (EG: Auth token stuff)
+  */
+/datum/db_query/proc/Execute(async = TRUE, log_error = TRUE)
+	Activity("Execute")
+	if(in_progress)
+		CRASH("Attempted to start a new query while waiting on the old one")
+
+	if(!SSdbcore.IsConnected())
```

Comment:
This does a rust call each time right? Isn't that a bit expensive

Reply:
Nope, since its an entirely RAM based lookup. FFIs arent expensive if handled properly (IE: We make a rust call every single time any log line is written, but its never noticed in round at all)